text/0000-enum-variant-types.md

```diff
@@ -0,0 +1,203 @@
+```
+
+# Motivation
+[motivation]: #motivation
+
+When working with enums, it is frequently the case that some branches of code have assurance that
+they are handling a particular variant of the enum. This is especially the case when abstracting
+behaviour for a certain enum variant. However, currently, this information is entirely hidden to the
+compiler and so the enum types must be matched upon even when the variant is certainly known.
+
+By treating enum variants as types in their own right, this kind of abstraction is made cleaner,
+avoiding the need for code patterns such as:
+- Passing a known variant to a function, matching on it, and use `unreachable!()` arms for the other
+variants.
+- Passing individual fields from the variant to a function.
+- Duplicating a variant as a standalone `struct`.
+
+However, though abstracting behaviour for specific variants is often convenient, it is understood
+that such variants are intended to be treated as enums in general. As such, the variant types
+proposed here have identical representations to their enums; the extra type information is simply
+used for type checking and permitting irrefutable matches on enum variants.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The variants of an enum are considered types in their own right, though they are necessarily
+more restricted than most user-defined types. This means that when one define an enum, one is more
+precisely defining a collection of types: the enumeration itself, as well as each of its
+variants. However, the variant types act very similarly to the enum type in the majority of cases.
+
+Specifically, variant types act differently to enum types in the following case:
+- When pattern-matching on a variant type, only the constructor corresponding to the variant is
+considered possible. Therefore one may irrefutably pattern-match on a variant.
+
+Variant types, unlike most user-defined types are subject to the following restriction:
+- Variant types may not have inherent impls, or implemented traits. That means `impl Enum::Variant`
+and `impl Trait for Enum::Variant` are forbidden. This dissuades inclinations to implement
+abstraction using behaviour-switching on enums, rather than using traits as is natural in Rust.
+
+Variant types may be aliased with type aliases.
+
+If a value of a variant type is explicitly cast to the type of its enum using a type annotation or
+by passing it as an argument or return-value to or from a function, the variant information is lost
+(that is, a variant type *is* different to an enum type, even though they behave very similarly).
+
+Note that enum types may not be coerced to variant types. Instead, matching must be performed to
+guarantee that the enum type truly is of the expected variant type.
+
+```rust
+enum Sum { A, B, C }
+
+let s: Sum = Sum::A;
+
+let a = s as Sum::A; // error
+let a: Sum::A = s; // error
+
+if let a @ Sum::A = s {
+    // ok, `a` has type `Sum::A`
```

Comment:
And may you use the fields of `Sum` with field-projection if `Sum::A` should have such?
If so, show that in an example, otherwise, why not?